### ISSUE TITLE
Auto-Regression fails when there is NaN in recent data

### ISSUE THREAD

**Issue by winedarksea** [ourownstory/neural_prophet#498]
```ValueError: need at least one array to concatenate```

Issue only occurs when there is NaN - I believe NaN data is meant to be handled by default, but if not, then you can ignore this Issue. 

### Code to Reproduce:
```python
import numpy as np
import pandas as pd
from neuralprophet import NeuralProphet
def load_sine(shape, start_date: str = "2021-01-01"):
    df_wide = pd.DataFrame(
        np.ones(shape),
        index=pd.date_range(start_date, periods=shape[0], freq="D"),
        columns=range(shape[1]),
    )
    X = pd.to_numeric(df_wide.index, errors='coerce', downcast='integer').values

    def sin_func(a, X):
        return a * np.sin(1 * X) + a

    for column in df_wide.columns:
        df_wide[column] = sin_func(column + 1, X)
    return df_wide

df = load_sine(shape=(300, 1)).reset_index(drop=False)
df.columns = ['ds', 'y']
df.iloc[-1, 1] = np.nan
forecast_length = 10

m = NeuralProphet(
    n_lags=forecast_length * 2,
    n_forecasts=forecast_length,
)

metrics = m.fit(df, freq='D')

df_future = m.make_future_dataframe(df, periods=forecast_length)
forecast = m.predict(df_future, decompose=False)
```

comment out this line and it works: ```df.iloc[-1, 1] = np.nan```

**Comment by winedarksea** [comment]
On a related note, `n_forecasts` only matters when `n_lags` is set, right?

**Comment by ourownstory** [comment]
Thank you @winedarksea for filing this bug, and sorry for not responding sooner.

I ran your code, and it seems you found a special edge case that produces a bug.
- Only occurs if `NaN` is at very end (not in middle or very beginning)
- Only if `make_future_dataframe` is called before, and
- Only if `make_future_dataframe` is called with `periods > 0` and `n_historic_predictions == 0 or False`

We will have a closer look at this.

Thank you,
Okar

**Comment by ourownstory** [comment]
Hi @winedarksea Yeah, I agree, it's a rather common edge case. We will update our tests to include this when we dive into a solution. 

Btw, would you mind sharing how you would expect missing values at the end to be filled in?

Options that come to mind:
minimal fix:
* last observed value repeated for all missing values

hard to generalize:
* average over last N, e.g. `n_lags` values (challenge: 'reasonable' length of past to take into account varies)
* some linear extrapolation (challenge: reasonable length of past to take into account varies)

most complicated / can produce inconsistent behavior:
* using the fitted model to predict  (challenge: if we have more missing values than n_forecasts, we may have to repeat the procedure several times.)

**Comment by winedarksea** [comment]
Even just improving the error message would probably be useful - "NaN found in recent data, not supported...", or similar.

As for filling NaN... a forward fill of the last observed value, with my data, is often the best in cross validation even while run against a number of fancier methods. And also being simple, I would vote for that.

As for  fancy methods, I have had good luck with using the [KNN Imputer](https://scikit-learn.org/stable/modules/generated/sklearn.impute.KNNImputer.html) and also the [akima interpolation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.Akima1DInterpolator.html#scipy.interpolate.Akima1DInterpolator). Although I am not sure they are worth the compute time when compared to the forward fill of last value.

**Comment by judussoari** [comment]
Hi @ourownstory, @winedarksea, 

I actually came across this problem while working on PR https://github.com/ourownstory/neural_prophet/pull/619, which has been merged into main recently.

While I did not discover this special edge case (`periods>0`, `n_historic_predictions=0` when creating `df_future`) while working on it, I realized `forecast` contains loads of NaN and None values, whenever `n_lags>0` and whenever there is at least one NaN value within `df_future` – or in other terms: within the last 
`x = n_lags`+`n_historic_predictions` entries of the df (be it at position `-1`,`-2`, or another one). To my understanding, this makes the forecast basically useless. 

My approach to 'solve' this was to simply throw a `ValueError: Data used for historic forecasts contains NaN values. Please ensure there are no NaN values within the last x entries of the df`. With this, we don't need to think about the edge case you discovered (because the error will be raised with `n_historic_predictions>=0`), and we encourage the user to have a look on their dataframe on top. Once they get rid of the NaN in `df_future`, `forecast` will be meaningful again. 

I like the idea of filling the NaN automatically instead of throwing an Error though. Would you also recommend a forward fill of the last observed value, even if the NaN is not at position `-1`? For example, if there is a NaN at position `-3`, would you say it should be filled with the value of position `-4`? 

Thanks :) 
Júlio

**Comment by winedarksea** [comment]
I am still unsure what fillna method is the best default. Ffill is fast and makes sense for most cases. I have the advantage in AutoTS of being able to try different approaches across cross validation, but a fairly common problem, where new data has more NaN's because it hasn't had time to be cleaned upstream in the source, isn't really dealt with well.

You can see all the fill na methods I have implemented, which I recently profiled and made faster:
https://github.com/winedarksea/AutoTS/blob/master/autots/tools/impute.py which also relies heavily on pandas/scipy interpolate. It's not the cleanest python script you'll ever see, but it's gathered basically ever fillna method that exists, including some of my own creation - fake date fill is a method where NaN values are thrown out then other values shifted forward, seems crazy but sometimes works quite well.

You could actually use AutoTS to search fillan methods since I've integrated Neural Prophet. Although current released version doesn't work with your current release (ar_sparsity, train_speed, progress_print got removed or something) but it's updated in dev. And also, I would welcome any suggestions or recommendations on improving the integration.

**Comment by ourownstory** [comment]
resolved in #653
